### PR TITLE
[Enhancement] The payload position should not be predictable

### PR DESCRIPTION
**Pull request by lcotonea**
The payload is now somewhere in the string instead of the first location candidate

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -73,23 +73,25 @@ function cliReveal(payload, password, op) {
...
+        const cleanedMessage = str.split(' ')[0] + (' ' + payload.slice(limit)) + str.split(' ').slice(2).join(' ')
+        if (invisible) {
+          return zwcStream
+        } else {
+          return cleanedMessage
+        }
+      }
+    }
+    throw new Error(
+      "Invisible stream not detected ! Please copy paste the stegcloak text sent by the sender"
```
Comment (nandahkrishna) [inline cli.js:93]:
```suggestion
      "Invisible stream not detected! Please copy and paste the StegCloak text sent by the sender."
```

File: components/message.js [hunk 2]
```diff
@@ -82,16 +82,19 @@ const zwcOperations = (zwc) => {
...
+      const intersected = intersection(zwc, zwcBound);
+      if (intersected.length !== 0) {
+        const limit = zwcBound.findIndex((x, i) => !~zwc.indexOf(x));
+        return eachWords[currentIndex].slice(0, limit);
+      }
     }
-    const limit = zwcBound.findIndex((x, i) => !~zwc.indexOf(x));
-    return payload.slice(0, limit);
+    throw new Error(
+      "Invisible stream not detected ! Please copy paste the stegcloak text sent by the sender"
```
Comment (nandahkrishna) [inline components/message.js:96]:
```suggestion
      "Invisible stream not detected! Please copy and paste the StegCloak text sent by the sender."
```

**Comment by JyothishmathiCV** [comment]
This discussion was something we were having for quite some time and therefore I thank @lcotonea for bringing it up so that we could get other's insight regarding the same. I will mention my views regarding this.

As stated by @nandahkrishna, the invisible stream must be present in the first few spaces and preferably the first one.

I would like to point out that the invisibility of StegCloak lies in the fact that people do not know the existence of such characters. This is achieved due to their invisible presence on the display.

However, if the presence of such characters is detected or considered, the position of the characters will not enhance their security. In fact if their existence is to be verified, It could easily be done so by using Ramda intersect function of the message and Stegcloak's array (like your Bruteforce check)  where the position of these characters will not increase the difficulty in their detection.

The reason we chose the first space was because it doesn't really matter which space it is present in as long as nobody knows of their presence since the detection can be a simple intersection check. Therefore it is convenient to fix the first space for embedding. 
 
So I would conclude it is not going to change anything, but if you have a valid justification please comment your thoughts :).

**Comment by lcotonea** [comment]
I thank you for your feedback.
I now understand the choice of the first character for the location of the payload.
I believe that steganography relies on discretion by using a public channel to expose encrypted information. It is true that the information is detectable, if you know where to look.
However I think that there are cases of use where the location of the payload must be random. The idea is not to strengthen the cache of information, but rather to free oneself from the constraint of the location of the ⁤⁣⁢‌⁣‌‍⁡⁢‍⁤⁡‌⁤⁡‍⁢⁣‌‍⁣‍‌⁢⁡‍‌‌⁡‍⁢⁣‍⁡⁤‍⁤⁡‍‍⁣‌⁢‌⁢‍⁤⁡‍‌‍⁡‍‌‌⁡‌⁡⁤‍⁤⁡‍‍⁣‍‌‍⁢‍⁤⁡‌⁡⁢⁡‍‍⁢⁡‍⁣⁢‍⁤⁡‍⁡‍⁡‍‍⁣‌⁢‍⁡‌⁢‌⁡‌⁡⁤‍⁤⁡‍‌⁢⁡‍‌‌⁡‍⁢‌⁡‍⁢⁣‌⁡⁢⁡‍‍⁣‍‌‌⁡‍‌‍⁢‍⁤⁢‌‍‍⁢‍‍⁡payload.
Example: let's say I receive several irc messages with a hidden message in one of them (I don't know which one). I'll copy the whole thing and run it through the reveal function. In this case, it is necessary not to look only at the first location where you expect to get the payload.
In conclusion, maybe the hide function can take an option to change the location of the payload (no use case in mind), so I leave it up to you. Nevertheless, I think the reveal function should be able to read the payload wherever it is in the text.

PS: an unencrypted payload is hidden in this text :)

**Comment by mohanpierce99** [comment]
@lcotonea  That makes sense. So what do you say if we keep the first space constraint but with Bruteforce check of the invisible stream instead of fixing it? By this, the reveal function would now be able to read the payload wherever it is in the text. Solving this use case of yours which I think is really good cause I've had similar problems at the same time maintaining the location.Kinda best of both worlds

**Comment by mohanpierce99** [comment]
I think this pr would be fine with the removal of changes in the embed function alone. @nandahkrishna   what do you think?

**Comment by nandahkrishna** [comment]
I think that there’s value in the random placement. It takes effort to discover where the secret is hidden (even if that’s easily done with a function, still is effort nonetheless). Currently no such effort is required.

@lcotonea in my opinion we just need some minor tweaks with respect to the bounds for placement, such as not hiding in the last space. Any thoughts on this, and maybe a nice heuristic that scales based on the cover text size?

**Comment by lcotonea** [comment]
> @lcotonea That makes sense. So what do you say if we keep the first space constraint but with Bruteforce check of the invisible stream instead of fixing it? By this, the reveal function would now be able to read the payload wherever it is in the text. Solving this use case of yours which I think is really good cause I've had similar problems at the same time maintaining the location.Kinda best of both worlds

For me it doesn't matter where. Besides, if I know that stegcloak encodes the payload in the first place, I'll know how to get the payload back.
As ⁤⁣⁢‍⁡⁢‍⁡‌⁡‍‌⁡⁢‍⁤⁡‍‌⁢⁡‍‍⁣‍‍‌⁡‍⁣⁢‍⁤⁡‍⁢⁡⁢‍⁤⁡‌⁢‌⁡‍⁣⁡‍⁢‌⁡‌⁢‍⁡‍⁣⁡‌⁡⁤‍⁤⁡‍⁢⁣‍⁡‌⁡‍⁣⁡‍‌‍⁡‌⁡⁤‍⁤⁡‌⁡‌⁡‍‍⁢⁡‍‌‌⁢‍⁤⁡‌⁢‌⁡‍⁣⁡‍‌‍⁡‍⁡⁤‍⁤⁡‍‌⁣‍⁣⁡‌⁢‌⁡‌⁢‌⁡‍⁢⁣‍⁡‌⁡‍⁣⁢‍⁤⁡‌⁡⁢⁡‍‍⁢⁡‌⁢‍⁡‍‌‌⁡‌⁣⁡‍⁡‌⁡‍‍⁤‍⁤⁡⁢‍⁣‍‌‍⁡‌⁡⁢⁡‍⁣⁡‌⁢‍⁡‍‌‍⁡‍⁣⁡‌⁡⁤‍⁤⁢‌‍‍⁢‍‍⁡soon as one knows the technique of steganography using UTF-8 hidden characters, it is easily detectable.
Finally, if stegcloak doesn't offer a reveal function that searches in any location, someone else will. The security of the "hiding place" cannot be based on the principle of secrecy.


> @lcotonea in my opinion we just need some minor tweaks with respect to the bounds for placement, such as not hiding in the last space. Any thoughts on this, and maybe a nice heuristic that scales based on the cover text size?

Yeah, that's a good idea. Can you give me more information on the heuristics? I'd like to better understand your idea.

**Comment by nandahkrishna** [comment]
If we have just about five words in a sentence (and hence 4 spaces), we could use the first 3. But as the text gets larger, say 50 words, we could avoid not just the last one space but the last few spaces. How does that sound?

I haven’t really thought about the math but this is just us trying to make it _slightly_ harder for a person to detect it. There are downsides to this too, so if we think that avoiding just the last place is better, then we’ll go ahead with that. Thoughts @mohanpierce99?

**Comment by JyothishmathiCV** [comment]
Maintaining a bound in such a way that the embedding will be within the first half of the text would be a good idea.

@lcotonea I think modifying the embed function to include this bound as follows would be great. 

`const targetIndex = Math.floor(Math.random() * Math.floor(arr.length/2));`

**Comment by mohanpierce99** [comment]
I'm cool with @JyothishmathiCV's change! Seems the right thing to do here @nandahkrishna Are you fine with this?

**Comment by lcotonea** [comment]
@JyothishmathiCV @mohanpierce99 @nandahkrishna I'm ok with the embed modification.

**Comment by nandahkrishna** [comment]
Same here, I’m okay with the changes proposed.

**Comment by lcotonea** [comment]
Thank to you.
And what about the reveal? 
> Example: let's say I receive several irc messages with a hidden message in one of them (I don't know which one). I'll copy the whole thing and run it through the reveal function. In this case, it is necessary not to look only at the first location where you expect to get the payload.

File: components/message.js [hunk 3]
```diff
@@ -82,16 +82,19 @@ const zwcOperations = (zwc) => {
...
   const detach = (str) => {
-    const payload = str.split(" ")[1];
-    const zwcBound = payload.split("");
-    const intersected = intersection(zwc, zwcBound);
-    if (intersected.length === 0) {
-      throw new Error(
-        "Invisible stream not detected ! Please copy paste the stegcloak text sent by the sender"
-      );
+    const eachWords = str.split(" ");
+    for(let currentIndex = 0; currentIndex < eachWords.length; currentIndex++) {
```
Comment (mohanpierce99) [inline components/message.js:86]: Please use forEach instead of for loops.

**Comment by mohanpierce99** [comment]
> Thank to you.
> And what about the reveal?
> 
> > Example: let's say I receive several irc messages with a hidden message in one of them (I don't know which one). I'll copy the whole thing and run it through the reveal function. In this case, it is necessary not to look only at the first location where you expect to get the payload.

Reveal has no problems tested the PR and the enhancement has been merged. Thank you for contributing :)
